drv/medusa-seq-server/src/front_io.rs

```diff
@@ -0,0 +1,111 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::*;
+use drv_fpga_api::{DeviceState, FpgaError};
+use drv_i2c_devices::{at24csw080::At24Csw080, Validate};
+use drv_sidecar_front_io::controller::FrontIOController;
+use drv_sidecar_front_io::phy_smi::PhySmi;
+
+#[allow(dead_code)]
+pub(crate) struct FrontIOBoard {
+    pub controllers: [FrontIOController; 2],
+    fpga_task: userlib::TaskId,
+    auxflash_task: userlib::TaskId,
+}
+
+impl FrontIOBoard {
+    pub fn new(
+        fpga_task: userlib::TaskId,
+        auxflash_task: userlib::TaskId,
+    ) -> Self {
+        Self {
+            controllers: [
+                FrontIOController::new(fpga_task, 0),
+                FrontIOController::new(fpga_task, 1),
+            ],
+            fpga_task,
+            auxflash_task,
+        }
+    }
+
+    pub fn phy(&self) -> PhySmi {
+        PhySmi::new(self.fpga_task)
+    }
+
+    pub fn present(i2c_task: userlib::TaskId) -> bool {
+        let fruid = i2c_config::devices::at24csw080_front_io0(i2c_task)[0];
+        At24Csw080::validate(&fruid).unwrap_or(false)
+    }
+
+    pub fn initialized(&self) -> bool {
+        self.controllers.iter().all(|c| c.ready().unwrap_or(false))
+    }
+
+    pub fn init(&mut self) -> Result<bool, FpgaError> {
+        let mut controllers_ready = true;
+
+        for (i, controller) in self.controllers.iter_mut().enumerate() {
+            let state = controller.await_fpga_ready(25)?;
+            let mut ident;
+            let mut ident_valid = false;
+            let mut checksum;
+            let mut checksum_valid = false;
+
+            if state == DeviceState::RunningUserDesign {
+                (ident, ident_valid) = controller.ident_valid()?;
+                ringbuf_entry!(Trace::FrontIOControllerIdent {
+                    fpga_id: i,
+                    ident
+                });
+
+                (checksum, checksum_valid) = controller.checksum_valid()?;
+                ringbuf_entry!(Trace::FrontIOControllerChecksum {
+                    fpga_id: i,
+                    checksum,
+                    expected: FrontIOController::short_checksum(),
+                });
+
+                if !ident_valid || !checksum_valid {
+                    // Attempt to correct the invalid IDENT by reloading the
+                    // bitstream.
+                    controller.fpga_reset()?;
+                }
+            }
+
+            if ident_valid && checksum_valid {
+                ringbuf_entry!(Trace::SkipLoadingFrontIOControllerBitstream {
+                    fpga_id: i
+                });
```

Comment:
it might be nice to have a comment here saying that we don't need to do anything because we are already running the expected bitstream. or, we could rename the ringbuf entry to something like `FrontIOControllerBitstreamAlreadyLoaded` so that it's clear to the reader why we stop here.

Reply:
As I state in the comment above, I'm going to leave this file as pure copy/paste as possible for now just to keep Sidecar/medusa as identical as possible until I can better unify the Front IO control stuff via #1805.